File: apps/ios/Packages/StoryArcEpub/Sources/EpubReaderFeature/ThemeSheet.swift
public import SwiftUI

internal import DesignSystem
internal import UIKit

public import StoryArcCore

/// The reading-theme sheet.
///
/// `ebook-reader` and `reading-themes` between them ask for a preset grid, a
/// stepped font size with a visible position, and — the part that is easy to skip —
/// an axis that cannot reach the page shown "unavailable with a one-line reason and
/// a single action that turns publisher styles off". Not hidden, and not a live
/// control that does nothing.
///
/// Custom backgrounds are Phase 3.7 and are not here yet. Everything else the
/// spec describes at both levels is.
struct ThemeSheet: View {
    @Environment(\.theme) var theme
    @Environment(\.accessibilityReduceMotion) var reduceMotion
    @Environment(\.dismiss) private var dismiss
    @Environment(\.dynamicTypeSize) private var typeSize

    let model: EpubReaderModel

    /// Words from where the reader is, read once when the sheet opens. Empty until the
    /// resource comes back, and empty for good on a publication it cannot be read from —
    /// the preview shows its sample paragraph in both cases.
    @State private var excerpt = ""

    /// Whether level two — the axes — is on screen.
    @State private var isCustomising = false

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: StoryArcSpace.xl) {
                    // First, because it is the thing every card below it previews.
                    ThemePreview(
                        readingTheme: model.theme,
                        values: model.values,
                        title: model.chapterTitle,
                        excerpt: excerpt
                    )
                    presets
                    customise
                }
                .padding(StoryArcSpace.gutter)
            }
            // Once, on open. The reader's position does not move while the sheet is up,
            // and re-reading the resource on every slider step would put a disk read
            // inside a drag.
            .task { excerpt = await model.previewExcerpt() }
            // No background of our own. A sheet on iOS 26 is already presented
            // on Liquid Glass, and `native-experience` wants it "left untinted so
            // it picks up the page beneath it" — an opaque fill here is the one
            // thing that would prevent that. The system's material also carries
            // its own Reduce-Transparency fallback, so declaring a second one
            // would only be able to disagree with it.
            // Inline, not a large title. A large title puts the sheet's name on its own
            // line under the toolbar, which costs a reader about 60 points of the page
            // they came here to adjust — on a sheet that is already only half the screen.
            .navigationTitle(Text("theme.title", bundle: .module))
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button { dismiss() } label: { Text("theme.done", bundle: .module) }
                }
            }
        }
        // Level two, as a second sheet from the first.
        //
        // Sheet-on-sheet is idiomatic on iOS and the platform animates it as a stack.
        // Android does not do this: `design.md` records why, and the short version is that
        // predictive back is a component-level contract there and two stacked modal sheets
        // give the gesture two competing dismiss targets and no correct preview.
        .sheet(isPresented: $isCustomising) {
            ThemeAxesSheet(model: model, excerpt: excerpt)
        }
    }

    /// The one action on level one, and the reason level one is only presets.
    ///
    /// `ebook-reader`: "one action, given equal prominence to the grid, opens the axes".
    /// Full-width and bordered-prominent, so it reads as the grid's peer rather than as a
    /// footnote under it — a reader who came to nudge line spacing has to be able to see
    /// where that lives without having learnt it first.
    private var customise: some View {
        Button { isCustomising = true } label: {
            Label {
                Text("theme.customise", bundle: .module)
            } icon: {
                Image(systemName: "slider.horizontal.3")
            }
            .frame(maxWidth: .infinity)
        }
        .buttonStyle(.borderedProminent)
        .controlSize(.large)
    }

    /// How many preset cards sit in a row at a given text size.
    ///
    /// A function rather than an expression inside the grid, so a host test can ask it. There
    /// is no simulator in this repository's unit loop, and the defect this encodes — *Original*
    /// wrapping to `Origi-` over `nal` at `AccessibilityXXXL` — is invisible to every test that
    /// reads source text rather than pixels. The arithmetic is assertable even where the
    /// rendering is not.
    static func presetColumns(for typeSize: DynamicTypeSize) -> Int {
        typeSize.isAccessibilitySize ? 1 : 3
    }

    /// Three by two, each card in its own colours — and one column at the accessibility sizes.
    ///
    /// `ebook-reader`: the grid previews "each preset in its own colours — six
    /// samples, not six labels". A swatch that took the app's palette would be six
    /// identical cards with different words on them.
    ///
    /// **The column count is not fixed, and it used to be.** Three columns leave a card about
    /// 170 pt wide on a 402 pt phone, which holds every preset's name at the ordinary text
    /// sizes and holds none of them at the accessibility ones: photographed at
    /// `AccessibilityXXXL` on 2026-09-05, *Original* wrapped mid-word and drew as `Origi-`
    /// over `nal`. A card is the one control in the app whose label may not shrink to fit —
    /// the whole point of the grid is that each name is drawn in its own typeface at its own
    /// weight, so shrinking it would be showing the reader the wrong typeface.
    ///
    /// One column rather than two at those sizes, because two still leaves ~170 pt and the
    /// name needs about 250 pt: two columns would move the wrap without preventing it. Six
    /// cards in one column is a longer scroll, which is the trade the accessibility sizes
    /// make everywhere else in this app.
    ///
    /// The neighbouring `PageColourSection` grids use `.adaptive(minimum:)` for the same job,
    /// and this one cannot: a swatch is square and interchangeable, so any number per row is
    /// as good as any other, while these cards must be *equal* width or the samples stop
    /// being comparable — which is what a fixed count gives and `.adaptive` does not.
    private var presets: some View {
        VStack(alignment: .leading, spacing: StoryArcSpace.sm) {
            Text("theme.presets", bundle: .module)
                .textRole(.headline)
                .foregroundStyle(theme.palette.textPrimary)
                // The trait, not `textRole`, is what lets VoiceOver jump section to section.
                .accessibilityAddTraits(.isHeader)

            LazyVGrid(
                columns: Array(
                    repeating: GridItem(spacing: StoryArcSpace.sm),
                    count: Self.presetColumns(for: typeSize)
                ),
                spacing: StoryArcSpace.sm
            ) {
                ForEach(ThemePreset.allCases, id: \.self) { preset in
                    PresetCard(
                        preset: preset,
                        isActive: model.theme.preset == preset && !model.theme.isCustom,
                        isModified: model.theme.preset == preset && model.theme.isModified
                    ) {
                        model.adopt(preset)
                        // `ebook-reader`: "picking a preset applies it and leaves the surface,
                        // because that was the whole errand". A sheet that stayed up over the
                        // change would put the reader's own decision behind their own choice.
                        dismiss()
                    }
                }
                // The seventh slot, present only once the reader has made one.
                // `reading-themes` puts it "alongside the six presets rather than
                // overwriting one", so it is a seventh card and not a replaced one.
                if let custom = model.theme.custom {
                    CustomCard(palette: custom, typeface: model.values.typeface) {
                        model.adoptColours(custom)
                    }
                }
            }
        }
    }

}

File: apps/android/feature/epubreader/src/main/kotlin/app/storyarc/feature/epubreader/ThemeSheet.kt
package app.storyarc.feature.epubreader

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.selection.toggleable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.TextDecrease
import androidx.compose.material.icons.filled.TextIncrease
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material.icons.filled.Tune
import androidx.compose.material3.Button
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.foundation.clickable
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material3.SheetState
import androidx.compose.material3.SheetValue
import androidx.compose.material3.rememberBottomSheetState
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import kotlinx.coroutines.launch
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Slider
import androidx.compose.material3.Surface
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import app.storyarc.core.designsystem.theme.LocalStoryArcPalette
import app.storyarc.core.designsystem.tokens.StoryArcRadius
import app.storyarc.core.designsystem.tokens.StoryArcSpace
import app.storyarc.core.model.FontSizeStep
import app.storyarc.core.model.PageTransition
import app.storyarc.core.model.ReaderPalette
import app.storyarc.core.model.ReaderTextAlignment
import app.storyarc.core.model.ReaderTypeface
import app.storyarc.core.model.ReadingTheme
import app.storyarc.core.model.STEPS_PER_AXIS
import app.storyarc.core.model.ThemeAxis
import app.storyarc.core.model.ThemePreset
import app.storyarc.core.model.ThemeValues
import app.storyarc.core.model.TransitionChoices
import app.storyarc.core.model.TransitionUnavailability
import app.storyarc.core.model.sliderRange
import app.storyarc.core.model.unit
import app.storyarc.core.model.value
import app.storyarc.core.model.values
import kotlin.math.roundToInt

/**
 * The reading-theme sheet.
 *
 * `ebook-reader` and `reading-themes` between them ask for a preset grid, a stepped
 * font size with a visible position, and — the part that is easy to skip — an axis
 * that cannot reach the page shown "unavailable with a one-line reason and a single
 * action that turns publisher styles off". Not hidden, and not a live control that
 * does nothing.
 *
 * The fine axes — line, character, word and paragraph spacing, margins, alignment,
 * custom background — are Phase 3.5 and 3.7 of the change and are not here yet.
 * What is here is the first level the spec describes. iOS's `ThemeSheet` is the same
 * sheet.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ThemeSheet(
    theme: ReadingTheme,
    values: ThemeValues,
    onAdopt: (ThemePreset) -> Unit,
    onAdoptColours: (ReaderPalette) -> Boolean,
    /** Opens level two: the axes, on a destination of their own. */
    onCustomise: () -> Unit,
    /**
     * The sheet this is drawn in, so the header row can toggle its height.
     *
     * Null in a preview, which has no sheet — the header then draws its name and nothing
     * else, which is what it did before Material asked for the toggle.
     */
    sheetState: SheetState? = null,
    modifier: Modifier = Modifier,
    /** The chapter the reader is in, for the live preview to name. */
    chapter: String? = null,
    /**
     * Words from where the reader is, read once when the sheet opens. Empty until the
     * resource comes back, and empty for good on a publication it cannot be read from --
     * the preview shows its sample paragraph in both cases.
     */
    excerpt: String = "",
) {
    val palette = LocalStoryArcPalette.current

    Column(
        modifier = modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(StoryArcSpace.gutter),
        verticalArrangement = Arrangement.spacedBy(StoryArcSpace.xl),
    ) {
        // First, because it is the thing every control below it changes.
        ThemePreview(theme = theme, values = values, title = chapter, excerpt = excerpt)

        HeightToggleHeader(sheetState)

        // Three by two, each card in its own colours. `ebook-reader`: the grid
        // previews "each preset in its own colours — six samples, not six labels".
        //
        // Rows rather than a `LazyVerticalGrid`: there are six or seven known items and
        // this is already inside a scrolling column, so a lazy grid buys nothing and
        // costs a *fixed height* — which at twice the system text size clipped the
        // labels off the bottom of every card. Rows take the height their content needs.
        val cards: List<@Composable (Modifier) -> Unit> = buildList {
            ThemePreset.entries.forEach { preset ->
                add { cardModifier ->
                    PresetCard(
                        preset = preset,
                        isActive = theme.preset == preset && !theme.isCustom,
                        isModified = theme.preset == preset && theme.isModified,
                        onSelect = { onAdopt(preset) },
                        modifier = cardModifier,
                    )
                }
            }
            // The seventh slot, present only once the reader has made one.
            // `reading-themes` puts it "alongside the six presets rather than
            // overwriting one", so it is a seventh card and not a replaced one.
            theme.custom?.let { custom ->
                add { cardModifier ->
                    CustomCard(
                        palette = custom,
                        typeface = values.typeface,
                        onSelect = { onAdoptColours(custom) },
                        modifier = cardModifier,
                    )
                }
            }
        }

        cards.chunked(COLUMNS).forEach { row ->
            Row(horizontalArrangement = Arrangement.spacedBy(StoryArcSpace.sm)) {
                row.forEach { card -> card(Modifier.weight(1f)) }
                // A short last row keeps its cards the same width as a full one.
                repeat(COLUMNS - row.size) { Spacer(Modifier.weight(1f)) }
            }
        }

        // The one action on level one, and the reason level one is only presets.
        //
        // `ebook-reader`: "one action, given equal prominence to the grid, opens the axes".
        // A filled button across the full width, so it reads as the grid's peer rather than
        // as a footnote under it — a reader who came to nudge line spacing has to be able to
        // see where that lives without having learnt it first.
        Button(
            onClick = onCustomise,
            modifier = Modifier.fillMaxWidth(),
        ) {
            Icon(
                imageVector = Icons.Filled.Tune,
                contentDescription = null,
                modifier = Modifier.padding(end = StoryArcSpace.xs),
            )
            Text(stringResource(R.string.theme_customise))
        }
    }
}

/**
 * The theme sheet, in the platform's own modal bottom sheet.
 *
 * `native-experience` wants the sheet to look like the platform's; iOS gets a
 * detented sheet on Liquid Glass and Android gets this.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ThemeBottomSheet(
    theme: ReadingTheme,
    values: ThemeValues,
    onAdopt: (ThemePreset) -> Unit,
    onAdoptColours: (ReaderPalette) -> Boolean,
    /** Opens level two: the axes, on a destination of their own. */
    onCustomise: () -> Unit,
    onDismiss: () -> Unit,
    chapter: String? = null,
    excerpt: String = "",
) {
    // `rememberModalBottomSheetState` is deprecated in 1.5.0-alpha26 with an exact
    // `replaceWith` pointing here: `rememberBottomSheetState(initialValue, enabledValues,
    // confirmValueChange)`. Verified by `javap` over `material3-1.5.0-alpha26.aar` rather than
    // trusted from the documentation.
    //
    // **`enabledValues` is now an explicit decision.** Since alpha21 the `PartiallyExpanded`
    // anchor is no longer removed for you, so a sheet that says nothing gets all three. Level
    // one wants all three: it opens at Material's 50% cap, expands to show the seventh card
    // and the action under the grid, and hides. That is what makes it multi-height, and what
    // makes the header toggle below something Material actually requires.
    val sheetState = rememberBottomSheetState(
        initialValue = SheetValue.PartiallyExpanded,
        enabledValues = setOf(SheetValue.Hidden, SheetValue.PartiallyExpanded, SheetValue.Expanded),
    )

    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = sheetState) {
        ThemeSheet(
            sheetState = sheetState,
            theme = theme,
            values = values,
            onAdopt = onAdopt,
            onAdoptColours = onAdoptColours,
            onCustomise = onCustomise,
            chapter = chapter,
            excerpt = excerpt,
        )
    }
}

/**
 * The sheet's name, and the single-pointer way to change its height.
 *
 * **This fills a gap that is Material's.** Material says "selecting the drag handle should
 * toggle through preset heights" and specifies a Space/Enter contract for it — and
 * `BottomSheetDefaults.DragHandle` has **no** `onClick`. There are zero `clickable` calls in
 * the whole sheet implementation, verified by reading it rather than assumed. So a
 * multi-height sheet owes a hand-built alternative that Material explicitly requires, and
 * level one is multi-height: it opens partially expanded and expands.
 *
 * The header row is where it goes. It is already the width of the sheet, it already carries
 * the sheet's name, and a whole row is well past the platform's minimum touch target — which
 * a 32×4dp handle is not. `Role.Button` with a state description is what gives a screen reader
 * the Space/Enter contract the handle was supposed to have.
 *
 * Delete this the day `DragHandle` takes an `onClick`.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun HeightToggleHeader(sheetState: SheetState?, modifier: Modifier = Modifier) {
    val palette = LocalStoryArcPalette.current
    val scope = rememberCoroutineScope()
    val isExpanded = sheetState?.currentValue == SheetValue.Expanded
    val action = stringResource(
        if (isExpanded) R.string.theme_sheet_collapse else R.string.theme_sheet_expand,
    )

    Row(
        modifier = modifier
            .fillMaxWidth()
            .then(
                if (sheetState == null) {
                    Modifier
                } else {
                    Modifier
                        .clickable {
                            scope.launch {
                                if (isExpanded) sheetState.partialExpand() else sheetState.expand()
                            }
                        }
                        .semantics {
                            role = Role.Button
                            stateDescription = action
                        }
                },
            ),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = stringResource(R.string.theme_presets),
            style = MaterialTheme.typography.titleMedium,
            color = palette.textPrimary,
            modifier = Modifier.weight(1f),
        )
        if (sheetState != null) {
            Icon(
                imageVector = if (isExpanded) {
                    Icons.Filled.KeyboardArrowDown
                } else {
                    Icons.Filled.KeyboardArrowUp
                },
                // The row carries the name and the state; a second announcement of the
                // same fact is one too many.
                contentDescription = null,
                tint = palette.textSecondary,
            )
        }
    }
}
